### ISSUE TITLE
HeidiSQL stalls while reading large UTF-8 MYSQL file after repeated "End of file block was cut within some multibyte character, at position xyz"

### ISSUE THREAD

**Issue by OlafEichstaedt** [HeidiSQL/HeidiSQL#354]
## Steps to reproduce this issue
1. create new/empty database X
2. select database X
3. select File/Run SQL File (File/Load SQL file will produce "Stream read error"
4. Encoding = UTF-8, Files of type = All files (*.*) 
5. select large MYSQL file created during a Drupal 7 backup
6. watch progress bar until HeidiSQL stalls at 20MiB/296MiB, 219'733 rows
7. cancel job to return to HeidiSQL

Note: Notepad++ or Sublime open the MYSQL-file without a problem and render the contained multibyte characters (e.g. arabic, chinese, french accents, ...) correctly.

## Current behavior
HeidiSQL seems to fail to read (large) MYSQL-files/queries that contain multi-byte characters. 

from line **20083** of logfile:
```
/* 2018-09-22 09:03:09 [] */ /* End of file block was cut within some multibyte character, at position 41'943'040. Increasing chunk size and retry reading... */
/* 2018-09-22 09:03:09 [] */ /* End of file block was cut within some multibyte character, at position 41'943'044. Increasing chunk size and retry reading... */
/* 2018-09-22 09:03:09 [] */ /* End of file block was cut within some multibyte character, at position 41'943'048. Increasing chunk size and retry reading... */
/* 2018-09-22 09:03:09 [] */ /* End of file block was cut within some multibyte character, at position 41'943'052. Increasing chunk size and retry reading... */
/* 2018-09-22 09:03:09 [] */ /* End of file block was cut within some multibyte character, at position 41'943'056. Increasing chunk size and retry reading... */
/* 2018-09-22 09:03:09 [] */ /* End of file block was cut within some multibyte character, at position 41'943'060. Increasing chunk size and retry reading... */
/* 2018-09-22 09:03:09 [] */ /* End of file block was cut within some multibyte character, at position 41'943'064. Increasing chunk size and retry reading... */
/* 2018-09-22 09:03:09 [] */ /* End of file block was cut within some multibyte character, at position 41'943'068. Increasing chunk size and retry reading... */
/* 2018-09-22 09:03:09 [] */ /* End of file block was cut within some multibyte character, at position 41'943'072. Increasing chunk size and retry reading... */
/* 2018-09-22 09:03:09 [] */ /* End of file block was cut within some multibyte character, at position 41'943'076. Increasing chunk size and retry reading... */
```
then from line  **20093** of logfile the messages repeat with the already mentioned  positions from 41'943'040 to 41'943'076, this continues until the last log entries from line **36232** of logfile:
```
/* 2018-09-22 09:09:32 [] */ /* End of file block was cut within some multibyte character, at position 41'943'076. Increasing chunk size and retry reading... */
/* 2018-09-22 09:19:51 [] */ /* Loading file "C:\[...]\SomeName.mysql" (296.6 MiB) into query tab #1 ... */
/* 2018-09-22 09:19:51 [] */ /* End of file block was cut within some multibyt
```
Note: the log file really ends with 'multibyt'.

## Expected behavior
HeidiSQL should be able to read large MYSQL file with UTF-8 encoded multibyte characters without problems.

## Possible solution
change strategy to read large files?

## Environment
* **HeidiSQL version: 9.5.0.5196** 
* **Database system and version: MariaDB 10.3.7** 
* **Operating system: Windows 10 Pro Version 10.0.17134 Build 17134**

**Comment by OlafEichstaedt** [comment]
Note: MySQL Workbench 8.0 CE connected to the same MariaDB excecuted the same MYSQL file without error.

**Comment by ansgarbecker** [comment]
Are you sure you selected "UTF-8" in the file-open-dialog?

Is there a chance I get that sql file, so I could reproduce this issue?

**Comment by OlafEichstaedt** [comment]
Hello Ansgar,

thanks for the quick response.

I just tried it again to be sure: yes UTF-8 is and was selected.

However, the SQL-File contains sensitive user data ... so I cannot send you the (whole) file.
But I am sure the section that causes the trouble with (too many?, an unlucky combination of?) multibyte characters is not sensitive ... so if you help me to understand the error log's "position" I could extract that section and send this part of the query to you. Maybe you can build a test file that pushes this part to a file block boundary and reproduce the undesired behavior. 

Or do you see any other way that I can help you narrow down the cause of this problem from my end?

**Comment by ansgarbecker** [comment]
That "position" from the log rows is the byte offset in the file. So, if you extract a portion from ~40MB up to ~43MB, that should be enough. You can also send it to anse at heidisql.com to keep it a bit more private than here in the tracker.

**Comment by OlafEichstaedt** [comment]
OK, let me see what I can do ... might need to defer it to tomorrow (Sunday) morning, though.

**Comment by OlafEichstaedt** [comment]
Hello Ansgar,

I sent you an email with a test case and log files.

After some research the root cause seems to be that there is non-UTF-8 content in the MySQL file.
It contains LONGBLOBs which have binary data inside.

So my question develops into:

* how to properly import/run such MySQL dumps?
* also: why does the huge MySQL file simply "hang" without a message and the shorter test case throws a "stream read error"?

**Comment by ansgarbecker** [comment]
Ok, I received your mail. And I looked at the .mysql file at first with Notepad++, which says it's ANSI encoded.

Importing it with HeidiSQL/"Run SQL file" with "UTF-8" selected in the file-open dialog, I get the same crash as you got. When using "ANSI" encoding, it does not crash, and the file seems to be imported correctly, although it may be due to the shortness of the file that it doesn't have UTF8 characters. Apart from the binary contents, there is just no other critical characters. Are you sure the other content is UTF8 encoded?

**Comment by OlafEichstaedt** [comment]
Hello Ansgar,

 

good question, I needed to double check.

 

I have tried to find out what Drupal7 says about their SQL export but did not find anything.

 

The MySQL script says 



SET NAMES utf8;

 

that – and the fact that we encode all the texts we import to Drupal in UTF-8 - send me down the UTF-8 path.

 

When I open the 300MB file in Notepad++ the editor says ANSI for me as well.

But the content does not render correctly, e.g. 

 

INSERT INTO `locales_source` VALUES 

('29948','/zh/stores','default','Cree(å…‹é‡Œè¯­)','','7.59'),

('29949','/zh/stores','default','Czech(æ·å…‹è¯­)','','7.59'),

('29952','/zh/stores','default','Welsh(å¨å°”å£«è¯­)','','7.59'),

('29953','/zh/stores','default','Danish(ä¸¹éº¦è¯­)','','7.59')

,('29954','/zh/stores','default','German(å¾·è¯­)','','7.59'),

[…]

('29976','/zh/stores','default','Guarani(ç“œæ‹‰å°¼è¯­)','','7.59'),

('29977','/zh/stores','default','ç‘žå£«å¾·è¯­','','7.59');

 

When I choose UTF-8 in Notepad++, I see:

 

INSERT INTO `locales_source` VALUES 

('29948','/zh/stores','default','Cree(克里语)','','7.59'),

('29949','/zh/stores','default','Czech(捷克语)','','7.59'),

('29952','/zh/stores','default','Welsh(威尔士语)','','7.59'),

('29953','/zh/stores','default','Danish(丹麦语)','','7.59'),

('29954','/zh/stores','default','German(德语)','','7.59'),

[…]

('29976','/zh/stores','default','Guarani(瓜拉尼语)','','7.59'),

('29977','/zh/stores','default','瑞士德语','','7.59');

 

… which are the correct Chinese translations for the names of different languages (the final character “话” meaning language). 

So at least this content seems to be UTF-8 encoded.

 

In HEX, Notepad++ gives me the following bytes

 

49 4e 53 45 52 54 20 49 4e 54 4f 20 60 6c 6f 63 61 6c 65 73 5f 73 6f 75 72 63 65 60 20 56 41 4c 55 45 53 20 

28 27 32 39 39 34 38 27 2c 27 2f 7a 68 2f 73 74 6f 72 65 73 27 2c 27 64 65 66 61 75 6c 74 27 2c 27 43 72 65 65 28 e5 85 8b e9 87 8c e8 af ad 29 27 2c 27 27 2c 27 37 2e 35 39 27 29 2c 

28 27 32 39 39 34 39 27 2c 27 2f 7a 68 2f 73 74 6f 72 65 73 27 2c 27 64 65 66 61 75 6c 74 27 2c 27 43 7a 65 63 68 28 e6 8d b7 e5 85 8b e8 af ad 29 27 2c 27 27 2c 27 37 2e 35 39 27 29 2c 

28 27 32 39 39 35 32 27 2c 27 2f 7a 68 2f 73 74 6f 72 65 73 27 2c 27 64 65 66 61 75 6c 74 27 2c 27 57 65 6c 73 68 28 e5 a8 81 e5 b0 94 e5 a3 ab e8 af ad 29 27 2c 27 27 2c 27 37 2e 35 39 27 29 2c 

28 27 32 39 39 35 33 27 2c 27 2f 7a 68 2f 73 74 6f 72 65 73 27 2c 27 64 65 66 61 75 6c 74 27 2c 27 44 61 6e 69 73 68 28 e4 b8 b9 e9 ba a6 e8 af ad 29 27 2c 27 27 2c 27 37 2e 35 39 27 29 2c 

28 27 32 39 39 35 34 27 2c 27 2f 7a 68 2f 73 74 6f 72 65 73 27 2c 27 64 65 66 61 75 6c 74 27 2c 27 47 65 72 6d 61 6e 28 e5 be b7 e8 af ad 29 27 2c 27 27 2c 27 37 2e 35 39 27 29 2c 

[…]

28 27 32 39 39 37 36 27 2c 27 2f 7a 68 2f 73 74 6f 72 65 73 27 2c 27 64 65 66 61 75 6c 74 27 2c 27 47 75 61 72 61 6e 69 28 e7 93 9c e6 8b 89 e5 b0 bc e8 af ad 29 27 2c 27 27 2c 27 37 2e 35 39 27 29 2c 

28 27 32 39 39 37 37 27 2c 27 2f 7a 68 2f 73 74 6f 72 65 73 27 2c 27 64 65 66 61 75 6c 74 27 2c 27 e7 91 9e e5 a3 ab e5 be b7 e8 af ad 27 2c 27 27 2c 27 37 2e 35 39 27 29 3b 0a  

 

which – according to http://www.ltg.ed.ac.uk/~richard/utf-8.cgi  are correct UTF-8 encoding.

 

So my conclusion is that this MySQL file also has UTF-8 encoded content. At least some of it.

 

I am a bit puzzled that the MySQL export from Drupal does render a file that does not have a single encoding … or is this the normal behaviour when we have strings and BLOBs in a database? 

And that – during import - an MySQL engine is supposed to read an un-encoded byte stream and choose an encoding based on the data type?

 

Best regards,

 

Olaf

**Comment by OlafEichstaedt** [comment]
Hello Ansgar,
 
good question, I needed to double check.

I have tried to find out what Drupal7 says about their SQL export but did not find anything.

The MySQL script says 

SET NAMES utf8;

that – and the fact that we encode all the texts we import to Drupal in UTF-8 - send me down the UTF-8 path.

When I open the 300MB file in Notepad++ the editor says ANSI for me as well.

But the content does not render correctly, e.g. 

```
INSERT INTO `locales_source` VALUES 
('29948','/zh/stores','default','Cree(å…‹é‡Œè¯­)','','7.59'),
('29949','/zh/stores','default','Czech(æ·å…‹è¯­)','','7.59'),
('29952','/zh/stores','default','Welsh(å¨å°”å£«è¯­)','','7.59'),
('29953','/zh/stores','default','Danish(ä¸¹éº¦è¯­)','','7.59'),
('29954','/zh/stores','default','German(å¾·è¯­)','','7.59'),
[…]
('29976','/zh/stores','default','Guarani(ç“œæ‹‰å°¼è¯­)','','7.59'),
('29977','/zh/stores','default','ç‘žå£«å¾·è¯­','','7.59');
```
 

When I choose UTF-8 in Notepad++, I see:
```
INSERT INTO `locales_source` VALUES 
('29948','/zh/stores','default','Cree(克里语)','','7.59'),
('29949','/zh/stores','default','Czech(捷克语)','','7.59'),
('29952','/zh/stores','default','Welsh(威尔士语)','','7.59'),
('29953','/zh/stores','default','Danish(丹麦语)','','7.59'),
('29954','/zh/stores','default','German(德语)','','7.59'),
[…]
('29976','/zh/stores','default','Guarani(瓜拉尼语)','','7.59'),
('29977','/zh/stores','default','瑞士德语','','7.59');
```
… which are the correct Chinese translations for the names of different languages (the final character “话” meaning language). 

So at least this content seems to be UTF-8 encoded.
In HEX, Notepad++ gives me the following bytes

```
49 4e 53 45 52 54 20 49 4e 54 4f 20 60 6c 6f 63 61 6c 65 73 5f 73 6f 75 72 63 65 60 20 56 41 4c 55 45 53 20 
28 27 32 39 39 34 38 27 [...] 28 **e5 85 8b e9 87 8c e8 af ad** 29 [...] 29 2c 
28 27 32 39 39 34 39 27[...] 28 **e6 8d b7 e5 85 8b e8 af ad** 29 [...]  29 2c 
28 27 32 39 39 35 32 27 2c 27 2f 7a 68 2f 73 74 6f 72 65 73 27 2c 27 64 65 66 61 75 6c 74 27 2c 27 57 65 6c 73 68 28 e5 a8 81 e5 b0 94 e5 a3 ab e8 af ad 29 27 2c 27 27 2c 27 37 2e 35 39 27 29 2c 
28 27 32 39 39 35 33 27 2c 27 2f 7a 68 2f 73 74 6f 72 65 73 27 2c 27 64 65 66 61 75 6c 74 27 2c 27 44 61 6e 69 73 68 28 e4 b8 b9 e9 ba a6 e8 af ad 29 27 2c 27 27 2c 27 37 2e 35 39 27 29 2c 
28 27 32 39 39 35 34 27 2c 27 2f 7a 68 2f 73 74 6f 72 65 73 27 2c 27 64 65 66 61 75 6c 74 27 2c 27 47 65 72 6d 61 6e 28 e5 be b7 e8 af ad 29 27 2c 27 27 2c 27 37 2e 35 39 27 29 2c 
[…]
28 27 32 39 39 37 36 27 2c 27 2f 7a 68 2f 73 74 6f 72 65 73 27 2c 27 64 65 66 61 75 6c 74 27 2c 27 47 75 61 72 61 6e 69 28 e7 93 9c e6 8b 89 e5 b0 bc e8 af ad 29 27 2c 27 27 2c 27 37 2e 35 39 27 29 2c 
28 27 32 39 39 37 37 27 2c 27 2f 7a 68 2f 73 74 6f 72 65 73 27 2c 27 64 65 66 61 75 6c 74 27 2c 27 e7 91 9e e5 a3 ab e5 be b7 e8 af ad 27 2c 27 27 2c 27 37 2e 35 39 27 29 3b 0a  
```
which – according to http://www.ltg.ed.ac.uk/~richard/utf-8.cgi  are correct UTF-8 encoding.

So my conclusion is that this MySQL file also has UTF-8 encoded content. At least some of it.

I am a bit puzzled that the MySQL export from Drupal does render a file that does not have a single encoding … or is this the normal behaviour when we have strings and BLOBs in a database? 

And that – during import - an MySQL engine is supposed to read an un-encoded byte stream and choose an encoding based on the data type?

Best regards,
Olaf

**Comment by ansgarbecker** [comment]
BLOB contents always have ASCII encoding, due to their binary nature. That's a problem when these come along with UTF-8 contents in the same file. Exporting binary content should be done in hexadecimal writing, which makes it compatible to all other encodings in the file.

In other words: the export has two different encodings: binary/ascii and utf-8. I suppose it's not possible to read such a file correctly, or?

**Comment by OlafEichstaedt** [comment]
MySQL workbench seems to change encoding between fields.
I guess this is how I managed to import/run the file **and** get UTF-8 and BLOB correctly.

When I manipulated the files in Python, I read the file as binary without encoding,
Then I separated the fields and applied the encoding as a function of data type. 

So I guess it is possible, but you cannot decide beforehand which encoding to use to read the whole file.
